predicates/interface_equivalent.py

```diff
@@ -1,78 +1,66 @@
+    CALLED_BY_NAME = auto()
+    DYNAMICALLY_SCOPED = auto()
+
+def ie_def(m: Macro, pd: PreprocessorData) -> IEResult:
     is_ = pd.mm[m]
     # We only analyze top-level non-argument invocations
     assert all([i.IsTopLevelNonArgument for i in is_])
     # We must have semantic data for all invocations
     if not all([i.HasSemanticData for i in is_]):
-        return False
+        return IEResult.NO_SEMANTIC_DATA
     # The macro must be expanded at least once
     if len(is_) == 0:
-        return False
+        return IEResult.MACRO_NEVER_EXPANDED
     # All invocations must have the same type signature
     if len(set([i.TypeSignature for i in is_])) != 1:
-        return False
+        return IEResult.INVOCATIONS_HAVE_DIFFERENT_TYPE_SIGNATURES
     # The macro must be defined at global scope
     if not m.IsDefinedAtGlobalScope:
-        return False
-    return (
-        (m.IsObjectLike and all([
-            all([
-                # Valid for analysis
-                i.HasSemanticData,
+        return IEResult.NOT_DEFINED_AT_GLOBAL_SCOPE
 
+    def check_conditions(i: Invocation):
+        CONDITIONS = [
+                # Valid for analysis
+                (i.HasSemanticData, IEResult.LACKS_SEMANTIC_DATA), 
                 # Can be turn into an enum or variable
-                i.IsObjectLike,
-                i.CanBeTurnedIntoEnumOrVariable,
-
+                (i.CanBeTurnedIntoEnumOrVariable if m.IsObjectLike else i.CanBeTurnedIntoFunction, IEResult.CANNOT_TRANSFORM),
                 # Argument-altering
-                not i.MustAlterArgumentsOrReturnTypeToTransform,
-
+                (not i.MustAlterArgumentsOrReturnTypeToTransform, IEResult.CALLED_BY_NAME),
                 # Declaration-altering
-                i.DefinitionLocationFilename not in pd.local_includes,
-                i.Name not in pd.inspected_macro_names,
-                not i.IsNamePresentInCPPConditional,
-                not i.MustAlterDeclarationsToTransform,
-
+                (i.DefinitionLocationFilename not in pd.local_includes, IEResult.DYNAMICALLY_SCOPED),
+                (i.Name not in pd.inspected_macro_names, IEResult.DYNAMICALLY_SCOPED),
+                (not i.IsNamePresentInCPPConditional, IEResult.DYNAMICALLY_SCOPED),
+                (not i.MustAlterDeclarationsToTransform, IEResult.DYNAMICALLY_SCOPED),
                 # Call-site-context-altering
-                not i.MustAlterCallSiteToTransform,
-
+                (not i.MustAlterCallSiteToTransform, IEResult.CALLSITE_CONTEXT_ALTERING),
                 # Thunkizing
-                not i.MustCreateThunksToTransform,
-
-                # Metaprogramming
-                not i.MustUseMetaprogrammingToTransform
-            ])
-            for i in is_
-        ])) or
-        (m.IsFunctionLike and all([
-            all([
-                # Valid for analysis
-                i.HasSemanticData,
+                (not i.MustCreateThunksToTransform, IEResult.THUNKIZING),
```

Comment:
On Monday let's talk about the predicates; the ones that we use in the MerC formalism are slightly different from the ones in the Maki paper and so we probably won't want to use the same groupings (e.g., thunkizing macros would fall under the "called by name" category in the MerC formalism). Let's discuss tomorrow in person so we can share our thoughts more quickly.

Reply:
I meant to shoot you a text saying I wanted to talk about some of the predicates - yeah discussing in person would be a good idea.